components/org.wso2.carbon.identity.application.authenticator.biometric/src/main/java/org/wso2/carbon/identity/application/authenticator/biometric/BiometricAuthenticator.java

```diff
@@ -0,0 +1,161 @@
+import static org.wso2.carbon.identity.application.authenticator.biometric.BiometricAuthenticatorConstants.CONTEXT_KEY;
+import static org.wso2.carbon.identity.application.authenticator.biometric.BiometricAuthenticatorConstants.DOMAIN_NAME;
+import static org.wso2.carbon.identity.application.authenticator.biometric.BiometricAuthenticatorConstants.WAIT_PAGE;
+
+/**
+ * Biometric Authenticator class.
+ */
+public class BiometricAuthenticator extends AbstractApplicationAuthenticator
+        implements FederatedApplicationAuthenticator {
+
+    private static final long serialVersionUID = 8272421416671799253L;
+    private static final Log log = LogFactory.getLog(BiometricAuthenticator.class);
+
+
+    @Override
+    public String getFriendlyName() {
+
+        return BiometricAuthenticatorConstants.AUTHENTICATOR_FRIENDLY_NAME;
+    }
+
+    @Override
+    public boolean canHandle(HttpServletRequest request) {
+        log.info("Signed challenge from the form submitted:" + request.getParameter(BiometricAuthenticatorConstants.
+                SIGNED_CHALLENGE));
+        return request.getParameter(BiometricAuthenticatorConstants.SIGNED_CHALLENGE) != null;
+    }
+
+    @Override
+    public String getContextIdentifier(javax.servlet.http.HttpServletRequest request) {
+
+        return request.getParameter(CONTEXT_KEY);
+    }
+
+    @Override
+    public String getName() {
+
+        return BiometricAuthenticatorConstants.AUTHENTICATOR_NAME;
+    }
+
+    @Override
+    protected void initiateAuthenticationRequest(HttpServletRequest request, HttpServletResponse response,
+                                                 AuthenticationContext context) throws AuthenticationFailedException {
+
+        AuthenticatedUser user = context.getSequenceConfig().getStepMap().
+                get(context.getCurrentStep() - 1).getAuthenticatedUser();
+        String username = user.getUserName();
+
+        Map<String, String> authenticatorProperties = context.getAuthenticatorProperties();
+        String serverKey = authenticatorProperties.get(BiometricAuthenticatorConstants.SERVER_KEY);
+        String fcmUrl = authenticatorProperties.get(BiometricAuthenticatorConstants.FCM_URL);
+
+        String hostname = IdentityUtil.getHostName();
+        String serviceProviderName = context.getServiceProviderName();
+        String message = username + " is trying to log into " + serviceProviderName + " from " + hostname;
+        // TODO: 2019-11-20  support localization-future improvement.Include in DOCs.
+        String sessionDataKey = request.getParameter(CONTEXT_KEY);
+
+        UUID challenge = UUID.randomUUID();
+        String randomChallenge = challenge.toString();
+        context.setProperty(BIOMETRIC_AUTH_CHALLENGE, randomChallenge);
+
+        DeviceDAOImpl biometricDAO = DeviceDAOImpl.getInstance();
+        String deviceID = biometricDAO.getDeviceID(username).get(0);
```

Comment:
We can change this to `String deviceID = DeviceDAOImpl.getInstance().getDeviceID(username).get(0);`